### PR TITLE
issue892 - added validation for constraints

### PR DESCRIPTION
**Pull request by LiilyZhang**
This PR is for first 2 items of issue 892 and unit test:
- Support version, integer, float, string, boolean and list of string types (Lily)
- Support boolean operators and aliases (&&, ||, etc) in the constraint language

### CODE REVIEW

File: externalpolicy/text_language/text_language.go [hunk 1]
```diff
@@ -20,9 +25,56 @@ func NewTextConstraintLanguagePlugin() plugin_registry.ConstraintLanguagePlugin
...
+	if !isConstraintExpression(dconstraints) {
+		return false, errors.New(fmt.Sprintf("The Constrain input: %v is not Contraint Express type", dconstraints))
+	}
 
 	// Validate that the expression is syntactically correct and parse-able
+	constraints = dconstraints.(externalpolicy.ConstraintExpression)
+
+	for _, constraint = range constraints {
+		// 1 constrain inside constrain list
+		fmt.Println("constraint: ", constraint)
```
Comment (dabooz) [inline externalpolicy/text_language/text_language.go]: No Printlns or Printfs...if you really want to log this then you need to use the glog.V(5).Infof(.........) API

File: externalpolicy/text_language/text_language.go [hunk 2]
```diff
@@ -69,3 +121,215 @@ func (p *TextConstraintLanguagePlugin) GetNextOperator(expression string) (strin
...
+
+	if !canParseToString(x) {
+		return false
+	}
+
+	if !strings.Contains(x, ",") {
+		return false
+	}
+
+	//s := strings.Split(x, ",")
```
Comment (dabooz) [inline externalpolicy/text_language/text_language.go]: remove this please

File: externalpolicy/text_language/text_language.go [hunk 3]
```diff
@@ -69,3 +121,215 @@ func (p *TextConstraintLanguagePlugin) GetNextOperator(expression string) (strin
...
+func isAllowedComparisonOpType(s string) bool {
+	if strings.Compare(s, lessthan) == 0 || strings.Compare(s, greaterthan) == 0 || strings.Compare(s, equalto) == 0 || strings.Compare(s, doubleequalto) == 0 || strings.Compare(s, lessthaneq) == 0 || strings.Compare(s, greaterthaneq) == 0 || strings.Compare(s, notequalto) == 0 || strings.Compare(s, inoperator) == 0 {
+		return true
+	}
+	return false
+}
+
+const andsimbol = "&&"
+const orsimbol = "||"
+const notsimbol = "^"
```
Comment (dabooz) [inline externalpolicy/text_language/text_language.go]: it is spelled 'symbol'

File: externalpolicy/text_language/text_language.go [hunk 4]
```diff
@@ -69,3 +121,215 @@ func (p *TextConstraintLanguagePlugin) GetNextOperator(expression string) (strin
...
+const and = "AND"
+const or = "OR"
+const not = "NOT"
+
+func isAllowedLogicalOpType(s string) bool {
+	if strings.Compare(s, andsimbol) == 0 || strings.Compare(s, orsimbol) == 0 || strings.Compare(s, notsimbol) == 0 || strings.Compare(s, and) == 0 || strings.Compare(s, or) == 0 || strings.Compare(s, not) == 0 {
+		return true
```
Comment (dabooz) [inline externalpolicy/text_language/text_language.go]: did you run 'make format' on this code?

File: externalpolicy/text_language/text_language_test.go [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+		"iame2edev == true && cpu == 3 NOT memory <= 32",
+		"hello == \"world\"",
+		"hello in \"'hi world', 'test'\"",
+		"eggs == \"truck load\" AND certification in \"USDA, Organic\"",
+		"version == 1.1.1 OR USDA == true",
+		"version in [1.1.1,INFINITY) ^ cert == USDA",
+	}
+	ce := externalpolicy.ConstraintExpression(constraintStrings)
+
+	fmt.Println("ce: ", ce)
```
Comment (dabooz) [inline externalpolicy/text_language/text_language_test.go]: within unit tests its better to use t.Log() than to use println

File: externalpolicy/text_language/text_language_test.go [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+	"testing"
+
+	"github.com/open-horizon/anax/externalpolicy"
+)
+
+func Test_Validate_Succeed1(t *testing.T) {
+	// boolean, int, string
+	textConstraintLanguagePlugin := NewTextConstraintLanguagePlugin()
+	constraintStrings := []string{
+		"iame2edev == true && cpu == 3 NOT memory <= 32",
```
Comment (dabooz) [inline externalpolicy/text_language/text_language_test.go]: question, so the NOT is treated as AND NOT?

File: externalpolicy/text_language/text_language_test.go [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+
+	var validated bool
+	var err error
+
+	validated, err = textConstraintLanguagePlugin.Validate(interface{}(ce))
+	if validated == true {
+		t.Errorf("Validation should fail but not, err: %v", err)
+	} else if err == nil {
+		t.Errorf("Validated should fail and return err, but didn't")
+	}
```
Comment (dabooz) [inline externalpolicy/text_language/text_language_test.go]: For the failure cases you should also check for the error text to make sure you get the right error.

File: externalpolicy/text_language/text_language.go (hunk 1 above)
Comment (LiilyZhang) [reply externalpolicy/text_language/text_language.go]: Thanks! This is most for debugging, will remove this line

File: externalpolicy/text_language/text_language.go (hunk 3 above)
Comment (LiilyZhang) [reply externalpolicy/text_language/text_language.go]: oops!

File: externalpolicy/text_language/text_language.go (hunk 4 above)
Comment (LiilyZhang) [reply externalpolicy/text_language/text_language.go]: 👌 will do that!

File: externalpolicy/text_language/text_language_test.go (hunk 6 above)
Comment (LiilyZhang) [reply externalpolicy/text_language/text_language_test.go]: Discussed over slack and we decided to remove `NOT` and `^`

File: externalpolicy/text_language/text_language.go [hunk 8]
```diff
@@ -69,3 +120,211 @@ func (p *TextConstraintLanguagePlugin) GetNextOperator(expression string) (strin
...
+// 6. for the version type, supported values are a single version or a range of versions in the semantic version format (the same as used for service verions). The == operator implies that the value is a single version. The 'in' operator treats the value as a version range. As with service versions, the version 1.0.0 when treated as a version range is equivalent to the explicit range [1.0.0,INFINITY).
+
+func validateOneConstraintExpression(expression string) (bool, error) {
+	if len(expression) == 0 {
+		return true, nil
+	}
+
+	pieces := strings.Split(expression, " ")
+	if len(pieces) < 3 {
+		return false, errors.New(fmt.Sprintf("found %v token(s), expecting 3 in an expression %v, expected form is <property> == <value> in constraint expression", len(pieces), expression))
```
Comment (dabooz) [inline externalpolicy/text_language/text_language.go]: expected form is <property> <operator> <value>

File: externalpolicy/text_language/text_language.go [hunk 9]
```diff
@@ -69,3 +120,211 @@ func (p *TextConstraintLanguagePlugin) GetNextOperator(expression string) (strin
...
+	}
+
+	compOp := pieces[1]
+	value := pieces[2]
+
+	// if will failed on case when string values that contain spaces but not quoted (starting from 2nd interation)
+	if !isAllowedComparisonOpType(pieces[1]) {
+		return false, errors.New(fmt.Sprintf("Expression: %v should contain valid comparison operator - wrong operator %v", expression, pieces[1]))
```
Comment (dabooz) [inline externalpolicy/text_language/text_language.go]: add the list of valid operators to the error message

File: externalpolicy/text_language/text_language.go [hunk 10]
```diff
@@ -69,3 +120,211 @@ func (p *TextConstraintLanguagePlugin) GetNextOperator(expression string) (strin
...
+	}
+
+	if canParseToFloat(value) || canParseToInteger(value) {
+		if strings.Compare(compOp, doubleequalto) == 0 || strings.Compare(compOp, equalto) == 0 || strings.Compare(compOp, lessthan) == 0 || strings.Compare(compOp, greaterthan) == 0 || strings.Compare(compOp, lessthaneq) == 0 || strings.Compare(compOp, greaterthaneq) == 0 {
+			return true, nil
+		}
+		return false, errors.New(fmt.Sprintf("1. Comparison operator: %v is not supported for numeric value: %v", compOp, value))
```
Comment (dabooz) [inline externalpolicy/text_language/text_language.go]: I think I understand why you included the prefix numbers in the error msg, e.g 1., 2. etc but we dont do that anywhere else in the code. It looks like there is enough context in the rest of msg to know the msg is coming from, I think you could remove the numbers.

File: externalpolicy/text_language/text_language.go (hunk 9 above)
Comment (dabooz) [reply externalpolicy/text_language/text_language.go]: It looks like you missed this one?

**Pull request merged**